policy/openbot/associate_frames.py

```diff
@@ -89,89 +93,143 @@ def associate(first_list, second_list, max_offset):
+            f.write(
+                "timestamp (frame),time_offset (cmd-frame),time_offset (ctrl-frame),frame,left,right,cmd\n"
+            )
+            for a, b in matches:
+                f.write(
+                    "%d,%d,%s,%s\n"
+                    % (a, b - a, ",".join(frame_list[a]), ",".join(cmd_list[b]))
+                )
         print(" Frames and commands matched.")
 
-    if (not redo_matching and os.path.isfile(os.path.join(sensor_path,"matched_frame_ctrl_cmd_processed.txt"))):
+    if not redo_matching and os.path.isfile(
+        os.path.join(sensor_path, "matched_frame_ctrl_cmd_processed.txt")
+    ):
         print(" Preprocessing already completed.")
     else:
-        #Cleanup: Add path and remove frames where vehicle was stationary
-        frame_list = read_file_list(os.path.join(sensor_path,"matched_frame_ctrl_cmd.txt"))
-        with open(os.path.join(sensor_path,"matched_frame_ctrl_cmd_processed.txt"), 'w') as f:
+        # Cleanup: Add path and remove frames where vehicle was stationary
+        frame_list = read_file_list(
+            os.path.join(sensor_path, "matched_frame_ctrl_cmd.txt")
+        )
+        with open(
+            os.path.join(sensor_path, "matched_frame_ctrl_cmd_processed.txt"), "w"
+        ) as f:
             f.write("timestamp,frame,left,right,cmd\n")
+            max_ctrl = get_max_ctrl(frame_list)
             for timestamp in list(frame_list):
-                if (len(frame_list[timestamp]) < 6):
+                frame = frame_list[timestamp]
+                if len(frame) < 6:
                     continue
-                left = int(frame_list[timestamp][3])
-                right = int(frame_list[timestamp][4])
-                if (remove_zeros and left-right==0 and left+right==0):
-                    print (" Removed timestamp:%s, left:%d, right:%d" %(timestamp,left,right))
-                    del frame_list[timestamp]
+                left = normalize(max_ctrl, frame[3])
+                right = normalize(max_ctrl, frame[4])
+                if remove_zeros and left == 0 and right == 0:
+                    print(f" Removed timestamp: {timestamp}")
+                    del frame
                 else:
-                    frame_name=(os.path.join(img_path,frame_list[timestamp][2]+"_crop.jpeg"))
-                    cmd = int(frame_list[timestamp][5])
-                    f.write("%s,%s,%d,%d,%d\n"%(timestamp,frame_name,left,right,cmd))
+                    frame_name = os.path.join(img_path, frame[2] + "_crop.jpeg")
+                    cmd = int(frame[5])
+                    f.write(
+                        "%s,%s,%d,%d,%d\n" % (timestamp, frame_name, left, right, cmd)
+                    )
         print(" Preprocessing completed.")
 
-    return read_file_list(os.path.join(sensor_path,"matched_frame_ctrl_cmd_processed.txt"))
+    return read_file_list(
+        os.path.join(sensor_path, "matched_frame_ctrl_cmd_processed.txt")
+    )
+
+
+def normalize(max_ctrl, val):
```

Comment:
I want to drive slowly to be exact.
Then the robot can go fast.
This can be an option. What do you think?

Reply:
Yes, this can be an option but should not be the default. The issue I still see is that the observations (images) would look different depending on speed and more aggressive control would be required. It is not necessarily a linear relationship. But would be interesting to try if it is possible to collect a dataset at a slow speed and then rescale.

Reply:
I commented it in the code.